src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -9629,64 +9620,29 @@ private ErrorPropertySymbol CreateErrorPropertySymbol(ImmutableArray<PropertySym
         }
 
         /// <summary>
-        /// Returns one of the methods from the method group if all methods in the method group
-        /// have the same signature, ignoring parameter names and custom modifiers. The particular
-        /// method returned is not important since the caller is interested in the signature only.
+        /// Returns one of the methods from the method group if all instance methods, or extension methods
+        /// in the nearest scope, have the same signature ignoring parameter names and custom modifiers.
+        /// The particular method returned is not important since the caller is interested in the signature only.
         /// </summary>
         private MethodSymbol? GetUniqueSignatureFromMethodGroup(BoundMethodGroup node)
         {
-            MethodSymbol? method = null;
-            foreach (var m in node.Methods)
-            {
-                switch (node.ReceiverOpt)
-                {
-                    case BoundTypeExpression:
-                    case null: // if `using static Class` is in effect, the receiver is missing
-                        if (!m.IsStatic) continue;
-                        break;
-                    case BoundThisReference { WasCompilerGenerated: true }:
-                        break;
-                    default:
-                        if (m.IsStatic) continue;
-                        break;
-                }
-                if (!isCandidateUnique(ref method, m))
-                {
-                    return null;
-                }
-            }
-            if (node.SearchExtensionMethods)
-            {
-                var receiver = node.ReceiverOpt!;
-                foreach (var scope in new ExtensionMethodScopes(this))
-                {
-                    var methodGroup = MethodGroup.GetInstance();
-                    PopulateExtensionMethodsFromSingleBinder(scope, methodGroup, node.Syntax, receiver, node.Name, node.TypeArgumentsOpt, BindingDiagnosticBag.Discarded);
-                    foreach (var m in methodGroup.Methods)
-                    {
-                        if (m.ReduceExtensionMethod(receiver.Type, Compilation) is { } reduced &&
-                            !isCandidateUnique(ref method, reduced))
-                        {
-                            methodGroup.Free();
-                            return null;
-                        }
-                    }
-                    methodGroup.Free();
-                }
-            }
+            var method = selectMethodForSignature(node);
+
             if (method is null)
             {
                 return null;
             }
-            int n = node.TypeArgumentsOpt.IsDefaultOrEmpty ? 0 : node.TypeArgumentsOpt.Length;
-            if (method.Arity != n)
+
+            int arity = node.TypeArgumentsOpt.IsDefaultOrEmpty ? 0 : node.TypeArgumentsOpt.Length;
+            if (method.Arity != arity)
             {
                 return null;
             }
```

Comment:
Are we inferring the delegate type for the following?
```csharp
var d = new C().M;

static class E
{
    public static void M<T>(this T t) { }
}
```

Reply:
Actually, it looks like this behavior didn't change. We weren't inferring the delegate type for this case previously either.

Reply:
> Actually, it looks like this behavior didn't change. We weren't inferring the delegate type for this case previously either.

Should we? Perhaps we should open an issue for that.

Reply:
Looked into it. `ReduceExtensionMethod` behaves like we want, but the final arity check (which Aleksey commented on, and exists in both the old and new logic) rules that answer out.
Instead of an arity check, we probably instead want to check that no type parameter was left unsubstituted.

Added this scenario to follow-up issue [https://github.com/dotnet/roslyn/issues/69222](https://github.com/dotnet/roslyn/issues/69222)